File: shared/contracts/signalr-events.ts
import type { JointPivotMappingOutput } from './joint-pivot-mapping-output';
import type { RawHardwareData } from './raw-hardware-data';

export const SignalREventName = {
  TelemetryJointState: 'telemetry.joint-state.updated',
  TelemetryJointAngleUpdate: 'telemetry.joint-angle.updated',
  TelemetryConnectionState: 'telemetry.connection.state',
  ControlSetJointTargets: 'control.set-joint-targets',
  ControlSetGripper: 'control.set-gripper',
} as const;

export type SignalREventName =
  (typeof SignalREventName)[keyof typeof SignalREventName];

export interface SignalREventEnvelope<TPayload> {
  eventName: SignalREventName;
  messageId: string;
  timestampUtc: string;
  source: 'hardware' | 'api' | 'web';
  payload: TPayload;
}

export interface TelemetryJointStatePayload {
  hardware: RawHardwareData;
  mapped: JointPivotMappingOutput;
}

export interface TelemetryJointAngleUpdatePayload {
  mapped: JointPivotMappingOutput;
}

export interface TelemetryConnectionStatePayload {
  isConnected: boolean;
  transport: 'serial' | 'wifi' | 'simulator';
  reason?: string;
}

export interface ControlSetJointTargetsPayload {
  hardwareTargets: RawHardwareData;
}

export interface ControlSetGripperPayload {
  openRatio: number;
}

File: shared/contracts/SignalREvents.cs
namespace RoboticV4.Contracts;

public static class SignalREventName
{
    public const string TelemetryJointState = "telemetry.joint-state.updated";
    public const string TelemetryConnectionState = "telemetry.connection.state";
    public const string ControlSetJointTargets = "control.set-joint-targets";
    public const string ControlSetGripper = "control.set-gripper";
}

public sealed record SignalREventEnvelope<TPayload>(
    string EventName,
    string MessageId,
    DateTimeOffset TimestampUtc,
    string Source,
    TPayload Payload
);

public sealed record TelemetryJointStatePayload(
    RawHardwareData Hardware,
    JointPivotMappingOutput Mapped
);

public sealed record TelemetryConnectionStatePayload(
    bool IsConnected,
    string Transport,
    string? Reason
);

public sealed record ControlSetJointTargetsPayload(
    RawHardwareData HardwareTargets
);

public sealed record ControlSetGripperPayload(
    double OpenRatio
);
